### PR TITLE
AI Component & System Implementation - Base

### PR DESCRIPTION
**Pull request by SeanAbnerNash**

### CODE REVIEW

File: ARGO/include/AiComponent.h [hunk 1]
```diff
@@ -1,12 +1,29 @@
 #pragma once
 #include "Component.h"
+#include "AIStates.h"
 
 class AiComponent : public Component
 {
 public:
+	AiComponent(AITypes t_types, AIStates t_states, float t_maxRot, float t_maxSpeed);
```
Comment (CathalRedmond) [inline ARGO/include/AiComponent.h]: the 'Rot' in maxRot is kinda ambiguous

Comment (CathalRedmond) [reply ARGO/include/AiComponent.h]: Same on line 16

File: ARGO/include/AiStates.h [hunk 2]
```diff
@@ -0,0 +1,3 @@
+#pragma once
+enum AIStates { eSleeping, eWander };
+enum AITypes { eMelee, eRanged };
```
Comment (CathalRedmond) [inline ARGO/include/AiStates.h]: 1). What is 'e', i assume enemy but its kinda ambiguous
2). Why enum and not enum class

File: ARGO/src/AiSystem.cpp [hunk 3]
```diff
@@ -15,8 +15,15 @@ void AiSystem::update(Entity& t_entity)
 		TransformComponent* posComp = static_cast<TransformComponent*>(t_entity.getAllComps().at(COMPONENT_ID::TRANSFORM_ID));
 		AiComponent* aiComp = static_cast<AiComponent*>(t_entity.getAllComps().at(COMPONENT_ID::AI_ID));
 
-		//switch statement to handle different ai types?
-		simpleMoveAi(posComp);
+		switch (aiComp->getType()) {
```
Comment (CathalRedmond) [inline ARGO/src/AiSystem.cpp]: move bracket down

File: ARGO/src/AiSystem.cpp [hunk 4]
```diff
@@ -15,8 +15,15 @@ void AiSystem::update(Entity& t_entity)
...
+		case eMelee:
+			meleeAI(posComp,aiComp);
+			break;
+		case eRanged:
+			rangedAI(posComp, aiComp);
```
Comment (CathalRedmond) [inline ARGO/src/AiSystem.cpp:24]: no break; here

File: ARGO/src/AiSystem.cpp [hunk 5]
```diff
@@ -15,8 +15,15 @@ void AiSystem::update(Entity& t_entity)
...
+		}
+		
+		//simpleMoveAi(posComp);
```
Comment (CathalRedmond) [inline ARGO/src/AiSystem.cpp]: is this needed?

File: ARGO/src/AiSystem.cpp (hunk 3 above)
Comment (CathalRedmond) [reply ARGO/src/AiSystem.cpp]: You copied this into melee ai and ranged ai  - has same errors there

File: ARGO/src/AiSystem.cpp (hunk 4 above)
Comment (CathalRedmond) [reply ARGO/src/AiSystem.cpp:24]: You copied this into melee ai and ranged ai  - has same errors there

File: ARGO/src/AiSystem.cpp [hunk 6]
```diff
@@ -58,3 +65,46 @@ void AiSystem::simpleMoveAi(TransformComponent* t_posComp)
...
+	tempAdjuster -= 10; //-10 to give a range of -9 and 10,
+	tempAdjuster /= 10;  //divide by 10 to give a range of -0.9 and 1
+	t_posComp->setRotation(t_posComp->getRotation() + (t_aiComponent->getMaxRotation() * tempAdjuster)); //Randomly changes current angle by the max which is scaled by the random scalar
+	glm::vec2 tempvelocity = glm::vec2(glm::cos(t_posComp->getRotation() * M_PI / 180), sin(t_posComp->getRotation() * M_PI / 180));//Generates a unit vector in the given angle.
+	tempvelocity *= glm::length(t_aiComponent->getMaxSpeed());//Scales it by the length of the max speed.
+	auto tempPosition = t_posComp->getPos();
+	tempPosition.x += tempvelocity.x;
+	tempPosition.y += tempvelocity.y;
+	t_posComp->setPos(tempPosition.x, tempPosition.y);//Updates Position
+}
```
Comment (CathalRedmond) [inline ARGO/src/AiSystem.cpp:114]: NO - FIX READABILITY

File: ARGO/src/AiSystem.cpp [hunk 7]
```diff
@@ -58,3 +65,46 @@ void AiSystem::simpleMoveAi(TransformComponent* t_posComp)
...
+		break;
+	case eWander:
+		wander(t_posComp, t_aiComponent);
+	}
+}
+
+void AiSystem::wander(TransformComponent* t_posComp, AiComponent* t_aiComponent)
+{
+	float tempAdjuster = (((rand() % 20 + 1)));//Gives a number between 1 and 20
+	tempAdjuster -= 10; //-10 to give a range of -9 and 10,
```
Comment (CathalRedmond) [inline ARGO/src/AiSystem.cpp]: why not just get random number between -9 and 10 instead of 1-20.
Also this can be done on the line above

Comment (CathalRedmond) [reply ARGO/src/AiSystem.cpp]: same on line below

File: ARGO/src/AiSystem.cpp [hunk 8]
```diff
@@ -58,3 +65,46 @@ void AiSystem::simpleMoveAi(TransformComponent* t_posComp)
...
+	tempAdjuster /= 10;  //divide by 10 to give a range of -0.9 and 1
+	t_posComp->setRotation(t_posComp->getRotation() + (t_aiComponent->getMaxRotation() * tempAdjuster)); //Randomly changes current angle by the max which is scaled by the random scalar
+	glm::vec2 tempvelocity = glm::vec2(glm::cos(t_posComp->getRotation() * M_PI / 180), sin(t_posComp->getRotation() * M_PI / 180));//Generates a unit vector in the given angle.
```
Comment (CathalRedmond) [inline ARGO/src/AiSystem.cpp]: there is no capital on velocity

Comment (CathalRedmond) [reply ARGO/src/AiSystem.cpp]: why use glm's cos function and c++'s sin function

File: ARGO/src/AiSystem.cpp [hunk 9]
```diff
@@ -58,3 +65,46 @@ void AiSystem::simpleMoveAi(TransformComponent* t_posComp)
...
+	tempvelocity *= glm::length(t_aiComponent->getMaxSpeed());//Scales it by the length of the max speed.
+	auto tempPosition = t_posComp->getPos();
+	tempPosition.x += tempvelocity.x;
+	tempPosition.y += tempvelocity.y;
+	t_posComp->setPos(tempPosition.x, tempPosition.y);//Updates Position
```
Comment (CathalRedmond) [inline ARGO/src/AiSystem.cpp]: You can do these 4 lines in 1:
t_posComp->setPos(t_posComp->getPos() + tempVelocity)

Comment (CathalRedmond) [reply ARGO/src/AiSystem.cpp]: if setPos takes only two floats, change to take/overload it to take a vector2

File: ARGO/src/TransformComponent.cpp [hunk 10]
```diff
@@ -80,4 +80,6 @@ void TransformComponent::moveUp()
 void TransformComponent::moveDown()
 {
 	m_pos.y++;
-}
\ No newline at end of file
+}
```
Comment (CathalRedmond) [inline ARGO/src/TransformComponent.cpp:83]: Why is the only change in this new lines at the bottom.
This doesn't change anything mate

**Review by CathalRedmond — CHANGES_REQUESTED** [review]

File: ARGO/include/AiStates.h (hunk 2 above)
Comment (SeanAbnerNash) [reply ARGO/include/AiStates.h]: 1 - Coding Convection for naming enum members.

File: ARGO/src/AiSystem.cpp (hunk 7 above)
Comment (SeanAbnerNash) [reply ARGO/src/AiSystem.cpp]: rand doesn't give negative values.

**Pull request merged**